pkg/comp-functions/functions/vshnpostgres/postgresql_deploy.go

```diff
@@ -2,31 +2,604 @@ package vshnpostgres
 
 import (
 	"context"
+	_ "embed"
+	"encoding/json"
 	"fmt"
+	"strings"
+	"time"
 
+	cmv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	certmgrv1 "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	xkubev1 "github.com/crossplane-contrib/provider-kubernetes/apis/object/v1alpha1"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
 	xfnproto "github.com/crossplane/function-sdk-go/proto/v1beta1"
+	promv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	sgv1 "github.com/vshn/appcat/v4/apis/stackgres/v1"
+	sgv1beta1 "github.com/vshn/appcat/v4/apis/stackgres/v1beta1"
+	appcatv1 "github.com/vshn/appcat/v4/apis/v1"
 	vshnv1 "github.com/vshn/appcat/v4/apis/vshn/v1"
+
+	"github.com/vshn/appcat/v4/pkg/common/utils"
 	"github.com/vshn/appcat/v4/pkg/comp-functions/functions/common"
 	"github.com/vshn/appcat/v4/pkg/comp-functions/runtime"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/ptr"
+)
+
+const (
+	certificateSecretName = "tls-certificate"
 )
 
+//go:embed scripts/copy-pg-backup.sh
+var postgresqlCopyJobScript string
+
 func DeployPostgreSQL(ctx context.Context, svc *runtime.ServiceRuntime) *xfnproto.Result {
 	l := svc.Log
 
 	comp := &vshnv1.VSHNPostgreSQL{}
 	err := svc.GetObservedComposite(comp)
 	if err != nil {
-		err = fmt.Errorf("cannot get observed composite: %w", err)
-		return runtime.NewFatalResult(err)
+		return runtime.NewFatalResult(fmt.Errorf("cannot get observed composite: %w", err))
 	}
 
 	l.Info("Bootstrapping instance namespace and rbac rules")
 	err = common.BootstrapInstanceNs(ctx, comp, "postgresql", "namespace-conditions", svc)
 	if err != nil {
-		err = fmt.Errorf("cannot bootstrap instance namespace: %w", err)
-		return runtime.NewWarningResult(err.Error())
+		return runtime.NewWarningResult(fmt.Errorf("cannot bootstrap instance namespace: %w", err).Error())
+	}
+
+	l.Info("Create tls certificate")
+	err = createCerts(comp, svc)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Errorf("cannot create tls certificate: %w", err).Error())
+	}
+
+	l.Info("Create Stackgres objects")
+	err = createStackgresObjects(ctx, comp, svc)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Errorf("cannot create stackgres objects: %w", err).Error())
+	}
+
+	l.Info("Create ObjectBucket")
```

Comment:
Thinking about this a bit more, it probably makes sense to create the bucket a lot earlier in the logic, as it stands now, the backups would also be destroyed.

Reply:
Good point. I moved it up before the instance namespace creation